src/scripting/lua_menu_item.cpp

```diff
@@ -0,0 +1,146 @@
+	//return_tstring_attrib("details", *((**cfg).details));  // optional
+	return_tstring_attrib("details", *(*ptr).details);  // optional
+	//return_string_attrib("icon", (**cfg).icon);
+	return_string_attrib("icon", (*ptr).icon);
+	//return_string_attrib("image", *(**cfg).image);  // optional
+	return_string_attrib("image", *(*ptr).image);  // optional
+	//return_tstring_attrib("label", (**cfg).label);
+*/
+	LOG_LUA << "impl_menuitem_get(): get label";
+	return_tstring_attrib("label", item.label);
+	//return_tstring_attrib("tooltip", (**cfg).tooltip);
+//	return_tstring_attrib("tooltip", (*ptr).tooltip);
+
+
+	return 0;
+}
+
+static int impl_menuitem_set(lua_State *L)
+{
+	gui2::menu_item& item = luaW_checkmenuitem(L, 1);
+
+	char const *m = luaL_checkstring(L, 2);
+
+//	if(lua_isnil(L, 3)) {
+	//	remove_config_attrib("checkbox", (**cfg));
+//		remove_config_attrib("details", (**cfg));
+//		remove_config_attrib("icon", (**cfg));
+//		remove_config_attrib("image", (**cfg));
+//		remove_config_attrib("label", (**cfg));
+//		remove_config_attrib("tooltip", (**cfg));
+//	} else {
+//		modify_bool_attrib("checkbox",  (**cfg)["checkbox"] = value);
+//		modify_bool_attrib("checkbox",  *(*ptr).checkbox = value);
+	LOG_LUA << "impl_menuitem_set(): set details";
+		modify_tstring_attrib("details", item.details = value);
+//		modify_string_attrib("icon",  (*ptr).icon = value);
+//		modify_string_attrib("image", *(*ptr).image = value);
+	LOG_LUA << "impl_menuitem_set(): set label";
+		//modify_tstring_attrib("label",  (**cfg).label = value);
+		modify_tstring_attrib("label",  item.label = value);
+//		modify_tstring_attrib("tooltip", (*ptr).tooltip = value);
+//	}
+	/*
+	if(lua_isnil(L, 3)) {
+		remove_config_attrib("checkbox", (**cfg));
+		remove_config_attrib("details", (**cfg));
+		remove_config_attrib("icon", (**cfg));
+		remove_config_attrib("image", (**cfg));
+		remove_config_attrib("label", (**cfg));
+		remove_config_attrib("tooltip", (**cfg));
+	} else {
+		modify_bool_attrib("checkbox",  (**cfg)["checkbox"] = value);
+		modify_tstring_attrib("details", (**cfg)["details"] = value);
+		modify_string_attrib("icon",  (**cfg)["icon"]= value);
+		modify_string_attrib("image", (**cfg)["image"] = value);
+		modify_tstring_attrib("label",  (**cfg)["label"] = value);
+		modify_tstring_attrib("tooltip", (**cfg)["tooltip"] = value);
+	}
+	*/
+
+	std::string err_msg = "invalid modifiable property of menu item: ";
+	err_msg += m;
+	return luaL_argerror(L, 2, err_msg.c_str());
+}
+
+namespace lua_menuitem {
+	std::string register_metatable(lua_State* L)
+	{
+		std::ostringstream cmd_out;
+
+		// Create the menu_item metatable.
+		cmd_out << "Adding menu_item metatable...\n";
+
+		luaL_newmetatable(L, "menu_item");
+		lua_pushcfunction(L, impl_menuitem_get);
+		lua_setfield(L, -2, "__index");
+		lua_pushcfunction(L, impl_menuitem_set);
+		lua_setfield(L, -2, "__newindex");
+		lua_pushstring(L, "menu_item");
+		lua_setfield(L, -2, "__metatable");
```

Comment:
Now that you're using `lua_ptr`, you will need to add a `__gc` metamethod. I think it should be sufficient to copy-and-paste almost any other `__gc` metamethod in the engine – most of them are identical except for the types involved.

Reply:
I looked at four, and they are pretty similar in that they call the objects destructor and return 0.

Currently, the menu_item doesn't have a destructor, and I'm not sure what it would do if it did.  And I'm not really sure why we would want to destroy the menu_item.  The only thing I could think of where gc would clean up the userdata would be something like if we defined the handle returned by add_item as local in preshow(), and then preshow finished and the handle went out of scope.  But in this case we don't want to destroy the object, it's still a valid row and we should still be able to refer to it by index, or get a new handle from get_row().

Assuming we did want to destroy the menu_item for some reason, we'd need to remove it from the vector?  I guess we could find its index by iterating over the vector and comparing the address of each row to the address of the ptr in our userdata?

I'm kind of lost here.

Reply:
> Currently, the menu_item doesn't have a destructor, and I'm not sure what it would do if it did.

Yes it does. All structs or classes have a destructor, even if you didn't declare one. It's not the menu_item that needs to be destroyed by the `__gc` metamethod though. It's the `lua_ptr<menu_item>`. You're not touching the menu item itself, only cleaning up that one reference to it.

It's _possible_ that the `lua_ptr` destructor _actually_ does nothing at all and we technically don't need this. However, we should have it anyway, just to be future-proof.

Reply:
Okay, now that makes sense.  ty